File: core/src/net/brenig/pixelescape/screen/MainMenuScreen.java
package net.brenig.pixelescape.screen;


import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.ui.*;
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener;
import com.badlogic.gdx.utils.Scaling;
import net.brenig.pixelescape.PixelEscape;
import net.brenig.pixelescape.game.data.GameMode;
import net.brenig.pixelescape.game.data.constants.StyleNames;
import net.brenig.pixelescape.lib.Reference;
import net.brenig.pixelescape.lib.Utils;
import net.brenig.pixelescape.render.ui.CurrentHighscoreLabel;
import net.brenig.pixelescape.render.ui.general.StageManager;
import net.brenig.pixelescape.render.ui.general.SwipeTabbedStack;

/**
 * PixelEscape MainMenu
 */
public class MainMenuScreen extends PixelScreen {

	private final StageManager uiStage;

	/**
	 * layout used to group main ui elements
	 */
	private final Table mainUiLayout;

	/**
	 * layout used to group setting buttons
	 */
	private final Table buttonPanelLayout;
	private final CurrentHighscoreLabel highscoreLabel;
	private final SwipeTabbedStack gmImageStack;

	public MainMenuScreen(final PixelEscape game) {
		super(game);
		//Setting up stage
		uiStage = new StageManager(game.getRenderManager());


		//Settings Button Panel
		buttonPanelLayout = Utils.createUIHeadLayout(game);
		//music and sound
		Utils.addSoundAndMusicControllerToLayout(game, buttonPanelLayout);
		//settings
		ImageButton btnSettings = new ImageButton(game.getSkin(), "settings");
		btnSettings.addListener(new ClickListener() {
			@Override
			public void clicked(InputEvent event, float x, float y) {
				game.setScreen(new SettingsScreen(game));
			}
		});
		btnSettings.getImageCell().fill().expand();
		buttonPanelLayout.add(btnSettings);
		buttonPanelLayout.invalidateHierarchy();
		//fullscreen
		Utils.addFullScreenButtonToTable(game, buttonPanelLayout);

		//Main UI Table
		mainUiLayout = new Table();
		mainUiLayout.setFillParent(true);
		mainUiLayout.setPosition(0, 0);
		mainUiLayout.center();

		//Center UI Table
		Table centerTable = new Table();

		//PixelEscape Heading
		Label header = new Label("PixelEscape", game.getSkin());
		header.setHeight(150);
		header.setFontScale(1.0F);

		centerTable.padTop(0);
		centerTable.add(header);
		centerTable.row();

		//GameMode Image
		gmImageStack = new SwipeTabbedStack(SwipeTabbedStack.DEFAULT_ANIMATION_X_OFFSET);
		//init gamemodes
		for (GameMode mode : game.getGameConfig().getAvailableGameModes()) {
			Image gameModeImage = new Image(mode.createIcon(game.getGameAssets()));
//			gameModeImage.setRotation(PixelEscape.rand.nextFloat() * 10 - 5F);
			gameModeImage.setScaling(Scaling.fit);
			gmImageStack.add(gameModeImage);
		}

		gmImageStack.setCurrentElement(game.getUserData().getLastGameMode());
		centerTable.add(gmImageStack).pad(20, 0, 10, 0).height(48).fillX();
		centerTable.row();

		//Highscore Label
		highscoreLabel = new CurrentHighscoreLabel(getGameMode());
		centerTable.add(highscoreLabel).padBottom(40);
		centerTable.row();

		//Buttons
		Table centerButtons = new Table();

		//Start Button
		TextButton btnStart = new TextButton("Start game", game.getSkin());
		btnStart.addListener(new ClickListener() {
			@Override
			public void clicked(InputEvent event, float x, float y) {
				game.setScreen(new GameScreen(game, getGameMode()));
			}
		});
		btnStart.getLabel().setFontScale(Reference.GAME_UI_MAIN_MENU_FONT_SIZE);
		centerButtons.add(btnStart).padBottom(8).fillX();

		//Quit Button
		if (game.getGameConfig().getCanQuitGame()) {
			TextButton btnQuit = new TextButton("Quit game", game.getSkin());
			btnQuit.addListener(new ClickListener() {
				@Override
				public void clicked(InputEvent event, float x, float y) {
					Gdx.app.exit();
				}
			});
			btnQuit.getLabel().setFontScale(Reference.GAME_UI_MAIN_MENU_FONT_SIZE);
			centerButtons.row();
			centerButtons.add(btnQuit).fillX();
		}
		centerTable.add(centerButtons);

		//Left Arrow
		Button arrow_left = new Button(game.getSkin(), StyleNames.BUTTON_ARROW_LEFT);
		arrow_left.addListener(new ClickListener() {
			@Override
			public void clicked(InputEvent event, float x, float y) {
				gmImageStack.last();
				highscoreLabel.setGameMode(getGameMode());
			}
		});
		mainUiLayout.add(arrow_left).size(96, 256);

		//Main UI
		mainUiLayout.add(centerTable);

		//Right Arrow
		Button arrow_right = new Button(game.getSkin(), StyleNames.BUTTON_ARROW_RIGHT);
		arrow_right.addListener(new ClickListener() {
			@Override
			public void clicked(InputEvent event, float x, float y) {
				gmImageStack.next();
				highscoreLabel.setGameMode(getGameMode());
			}
		});
		mainUiLayout.add(arrow_right).size(96, 256);

		//Move Arrows to front
		arrow_left.toFront();
		arrow_right.toFront();

		//Add ui elements to stage
		uiStage.getRootTable().top().right().pad(4);
		uiStage.add(buttonPanelLayout);
		uiStage.addActorToStage(mainUiLayout);
	}

	private GameMode getGameMode() {
		if (gmImageStack == null) {
			throw new IllegalStateException("GameMode select UI not initialized!");
		}
		return game.getGameConfig().getAvailableGameModes().get(gmImageStack.getCurrentElement());
	}

	@Override
	public void show() {
		game.getRenderManager().resetFontSize();
		uiStage.updateViewportToScreen();
		mainUiLayout.invalidateHierarchy();
		buttonPanelLayout.invalidateHierarchy();
		game.getGameMusic().playOrFadeInto(game.getGameAssets().getMainMenuMusic());
		Gdx.input.setInputProcessor(uiStage.getInputProcessor());
	}

	@Override
	public void render(float delta) {
		uiStage.act(delta);
		uiStage.draw(game.getRenderManager());
	}

	@Override
	public void resize(int width, int height) {
		uiStage.updateViewport(width, height, true);
		mainUiLayout.invalidateHierarchy();
	}

	@Override
	public void pause() {

	}

	@Override
	public void resume() {

	}

	@Override
	public void hide() {
		dispose();
	}

	@Override
	public void dispose() {
		game.getUserData().setLastGameMode(gmImageStack.getCurrentElement());
		uiStage.dispose();
	}
}

File: core/src/net/brenig/pixelescape/screen/MainMenuScreen.kt
package net.brenig.pixelescape.screen


import com.badlogic.gdx.Gdx
import com.badlogic.gdx.scenes.scene2d.InputEvent
import com.badlogic.gdx.scenes.scene2d.ui.*
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener
import com.badlogic.gdx.utils.Scaling
import net.brenig.pixelescape.PixelEscape
import net.brenig.pixelescape.game.data.GameMode
import net.brenig.pixelescape.game.data.constants.Reference
import net.brenig.pixelescape.game.data.constants.StyleNames
import net.brenig.pixelescape.lib.utils.UiUtils
import net.brenig.pixelescape.render.ui.CurrentHighscoreLabel
import net.brenig.pixelescape.render.ui.general.DisabledTextTooltip
import net.brenig.pixelescape.render.ui.general.HorizontalSpacer
import net.brenig.pixelescape.render.ui.general.PlayServiceLoginButton
import net.brenig.pixelescape.render.ui.general.SwipeTabbedStack

/**
 * PixelEscape MainMenu
 */
class MainMenuScreen(game: PixelEscape) : ScreenWithUi(game) {

    /**
     * layout used to group main ui elements
     */
    private val mainUiLayout: Table

    /**
     * layout used to group setting buttons
     */
    private val buttonPanelLayout: Table = UiUtils.createUIHeadLayout(game)
    private val highscoreLabel: CurrentHighscoreLabel
    private val gmImageStack: SwipeTabbedStack?
    private val playServiceButton: PlayServiceLoginButton?
    private val btnLeaderboards: ImageTextButton?

    private val gameMode: GameMode
        get() {
            if (gmImageStack == null) {
                throw IllegalStateException("GameMode select UI not initialized!")
            }
            return game.gameConfig.availableGameModes[gmImageStack.currentElement]
        }

    init {
        //Settings Button Panel
        //music and sound
        UiUtils.addSoundAndMusicControllerToLayout(game, buttonPanelLayout)
        //settings
        val btnSettings = ImageButton(game.skin, "settings")
        btnSettings.addListener(object : ClickListener() {
            override fun clicked(event: InputEvent, x: Float, y: Float) {
                game.screen = SettingsScreen(game)
            }
        })
        btnSettings.imageCell.fill().expand()
        buttonPanelLayout.add(btnSettings)
        buttonPanelLayout.invalidateHierarchy()
        //fullscreen
        UiUtils.addFullScreenButtonToTable(game, buttonPanelLayout)

        //Main UI Table
        mainUiLayout = Table()
        mainUiLayout.setFillParent(true)
        mainUiLayout.setPosition(0f, 0f)
        mainUiLayout.center()

        //Center UI Table
        val centerTable = Table()

        //PixelEscape Heading
        val header = Label("PixelEscape", game.skin)
        header.height = 150f
        header.setFontScale(1.0f)

        centerTable.padTop(0f)
        centerTable.add(header)
        centerTable.row()

        //GameMode Image
        gmImageStack = SwipeTabbedStack(SwipeTabbedStack.DEFAULT_ANIMATION_X_OFFSET)
        //init gamemodes
        for (mode in game.gameConfig.availableGameModes) {
            val gameModeImage = Image(mode.createIcon(game.gameAssets))
            //			gameModeImage.setRotation(PixelEscape.rand.nextFloat() * 10 - 5F);
            gameModeImage.setScaling(Scaling.fit)
            gmImageStack.add(gameModeImage)
        }

        gmImageStack.replaceCurrentELement(game.userData.lastGameMode)
        centerTable.add(gmImageStack).pad(20f, 0f, 10f, 0f).height(48f).fillX()
        centerTable.row()

        //Highscore Label
        highscoreLabel = CurrentHighscoreLabel(gameMode)
        centerTable.add(highscoreLabel).padBottom(40f)
        centerTable.row()

        //Buttons
        val centerButtons = Table()

        //Start Button
        val btnStart = TextButton("Start game", game.skin)
        btnStart.addListener(object : ClickListener() {
            override fun clicked(event: InputEvent, x: Float, y: Float) {
                game.screen = GameScreen(game, gameMode)
            }
        })
        btnStart.label.setFontScale(Reference.GAME_UI_MAIN_MENU_FONT_SIZE)
        centerButtons.add(btnStart).padBottom(8f).fillX()

        if (game.gameConfig.gameServiceAvailable) {
            btnLeaderboards = ImageTextButton("Leaderboards", game.skin, StyleNames.LEADERBOARDS)
            with(btnLeaderboards) {
                addListener(object : ClickListener() {
                    override fun clicked(event: InputEvent?, x: Float, y: Float) {
                        game.gameConfig.gameService.showLeaderboards(gameMode.scoreboardName)
                    }
                })
                val tooltip = DisabledTextTooltip("Login to Play Services to view leaderboards", game.skin)
                tooltip.setInstant(true)
                tooltip.actor.setFontScale(Reference.GAME_UI_SMALL_FONT_SIZE)
                tooltip.container.pad(4F)
                addListener(tooltip)
                label.setFontScale(Reference.GAME_UI_MAIN_MENU_FONT_SIZE)
                isDisabled = !game.gameConfig.gameService.isSessionActive
                pad(8F)
                image.setScaling(Scaling.fit)
                imageCell.size(UiUtils.buttonSize)
                imageCell.fill()
                image.setSize(UiUtils.buttonSize, UiUtils.buttonSize)
            }
            centerButtons.row()
            centerButtons.add(btnLeaderboards).padBottom(8f).fillX()
        } else {
            btnLeaderboards = null
        }

        //Quit Button
        if (game.gameConfig.canQuitGame) {
            val btnQuit = TextButton("Quit game", game.skin)
            btnQuit.addListener(object : ClickListener() {
                override fun clicked(event: InputEvent, x: Float, y: Float) {
                    Gdx.app.exit()
                }
            })
            btnQuit.label.setFontScale(Reference.GAME_UI_MAIN_MENU_FONT_SIZE)
            centerButtons.row()
            centerButtons.add(btnQuit).fillX()
        }
        centerTable.add(centerButtons)

        //Left Arrow
        val arrowLeft = Button(game.skin, StyleNames.BUTTON_ARROW_LEFT)
        arrowLeft.addListener(object : ClickListener() {
            override fun clicked(event: InputEvent, x: Float, y: Float) {
                gmImageStack.last()
                highscoreLabel.setGameMode(gameMode)
            }
        })
        mainUiLayout.add(arrowLeft).size(96f, 256f)

        //Main UI
        mainUiLayout.add(centerTable)

        //Right Arrow
        val arrowRight = Button(game.skin, StyleNames.BUTTON_ARROW_RIGHT)
        arrowRight.addListener(object : ClickListener() {
            override fun clicked(event: InputEvent, x: Float, y: Float) {
                gmImageStack.next()
                highscoreLabel.setGameMode(gameMode)
            }
        })
        mainUiLayout.add(arrowRight).size(96f, 256f)

        //Move Arrows to front
        arrowLeft.toFront()
        arrowRight.toFront()

        //Add ui elements to stage
        uiStage.rootTable.top().left().pad(4f)

        if (game.gameConfig.gameServiceAvailable) {
            playServiceButton = PlayServiceLoginButton(game.skin, StyleNames.SERVICE_LOGIN, StyleNames.SERVICE_WORKING, StyleNames.SERVICE_LOGOUT, game.gameConfig.gameService)
            playServiceButton.addListener(object : ClickListener() {
                override fun clicked(event: InputEvent?, x: Float, y: Float) {
                    btnLeaderboards?.isDisabled = !game.gameConfig.gameService.isSessionActive
                }
            })
            uiStage.add(playServiceButton)
        } else {
            playServiceButton = null
        }

        uiStage.add(HorizontalSpacer())
        uiStage.add(buttonPanelLayout)
        uiStage.addActorToStage(mainUiLayout)
    }

    override fun gsOnSessionActive() {
        super.gsOnSessionActive()
        playServiceStateUpdate()
    }

    override fun gsOnSessionInactive() {
        super.gsOnSessionInactive()
        playServiceStateUpdate()
    }

    private fun playServiceStateUpdate() {
        playServiceButton?.updateInfo()
        btnLeaderboards?.isDisabled = !game.gameConfig.gameService.isSessionActive
    }

    override fun show() {
        game.renderManager.resetFontSize()
        uiStage.updateViewportToScreen()
        mainUiLayout.invalidateHierarchy()
        buttonPanelLayout.invalidateHierarchy()
        @Suppress("ConstantConditionIf")
        if (Reference.ENABLE_MUSIC) game.gameMusic.playOrFadeInto(game.gameAssets.mainMenuMusic)
        Gdx.input.inputProcessor = uiStage.inputProcessor
    }

    override fun render(delta: Float) {
        uiStage.act(delta)
        uiStage.draw(game.renderManager)
    }

    override fun resize(width: Int, height: Int) {
        uiStage.updateViewport(width, height, true)
        mainUiLayout.invalidateHierarchy()
    }

    override fun pause() {

    }

    override fun resume() {

    }

    override fun hide() {
        dispose()
    }

    override fun dispose() {
        game.userData.lastGameMode = gmImageStack!!.currentElement
        uiStage.dispose()
    }
}
